spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/client/AbstractTraceHttpRequestInterceptor.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright 2013-2016 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.sleuth.instrument.web.client;
+
+import org.springframework.cloud.sleuth.Span;
+import org.springframework.cloud.sleuth.SpanAccessor;
+import org.springframework.cloud.sleuth.event.ClientReceivedEvent;
+import org.springframework.context.ApplicationEvent;
+import org.springframework.context.ApplicationEventPublisher;
+import org.springframework.context.ApplicationEventPublisherAware;
+import org.springframework.http.HttpRequest;
+import org.springframework.util.StringUtils;
+
+/**
+ * Abstraction over classes that interact with Http requests. Allows you
+ * to enrich the request headers with trace related information.
+ *
+ * @author Marcin Grzejszczak
+ */
+public abstract class AbstractTraceHttpRequestInterceptor
+		implements ApplicationEventPublisherAware {
+
+	private ApplicationEventPublisher publisher;
+	private final SpanAccessor accessor;
+
+	public AbstractTraceHttpRequestInterceptor(SpanAccessor accessor) {
+		this.accessor = accessor;
+	}
+
+	@Override
+	public void setApplicationEventPublisher(ApplicationEventPublisher publisher) {
+		this.publisher = publisher;
+	}
+
+	/**
+	 * Adds trace related headers from the span to the request
+	 */
+	public void enrichWithTraceHeaders(HttpRequest request, Span span) {
+		setHeader(request, Span.TRACE_ID_NAME, span.getTraceId());
+		setHeader(request, Span.SPAN_ID_NAME, span.getSpanId());
+		if (!span.isExportable()) {
+			setHeader(request, Span.NOT_SAMPLED_NAME, "true");
+		}
+		setHeader(request, Span.SPAN_NAME_NAME, span.getName().toString());
+		setHeader(request, Span.PARENT_ID_NAME, getParentId(span));
+		setHeader(request, Span.PROCESS_ID_NAME, span.getProcessId());
+	}
+
+	private Long getParentId(Span span) {
+		return !span.getParents().isEmpty() ? span.getParents().get(0) : null;
+	}
+
+	public void setHeader(HttpRequest request, String name, String value) {
```

Comment:
pls add javadoc for public non-internal symbols (ps this encourages less visibility :P)